src/main/java/tutorlink/lists/GradeList.java

```diff
@@ -0,0 +1,104 @@
+        componentDescription = componentDescription.toLowerCase();
+        for (Grade grade : gradeArrayList) {
+            if (grade.getStudent().getMatricNumber().equals(matricNumber)
+                    && grade.getComponent().getName().equals(componentDescription)) {
+                return gradeArrayList.remove(grade);
+            }
+        }
+        return false;
+    }
+
+    public void addGrade(Grade grade)
+            throws DuplicateGradeException {
+
+        for (Grade gradeToCompare : gradeArrayList) {
+            if (grade.equals(gradeToCompare)) {
+                throw new DuplicateGradeException(ERROR_DUPLICATE_GRADE_ON_ADD);
+            }
+        }
+        gradeArrayList.add(grade);
+    }
+
+    @Override
+    public String toString() {
+        return IntStream.range(0, gradeArrayList.size())
+                .mapToObj(i -> (i + 1) + ": " + gradeArrayList.get(i)) // 1-based index
+                .collect(Collectors.joining("\n\t"));
+    }
+
+    public GradeList findGrade(String matricNumber, String componentDescription) throws TutorLinkException {
+        GradeList filteredList = new GradeList();
+        filteredList.gradeArrayList = gradeArrayList
+                .stream()
+                .filter(grade -> grade.getStudent().getMatricNumber().equals(matricNumber.toUpperCase())
+                        && grade.getComponent().getName().equals(componentDescription.toLowerCase()))
+                .collect(Collectors.toCollection(ArrayList::new));
+        if (filteredList.gradeArrayList.isEmpty()) {
+            throw new GradeNotFoundException(String.format(ERROR_NO_GRADE_FOUND, matricNumber, componentDescription));
+        }
+        return filteredList;
+    }
+
+    public double calculateStudentGPA(String matricNumber, ComponentList componentList)
+            throws IncompleteGradesException {
+        ArrayList<Grade> studentGrades = gradeArrayList
+                .stream()
+                .filter(grade -> grade.getStudent().getMatricNumber().equals(matricNumber.toUpperCase()))
+                .collect(Collectors.toCollection(ArrayList::new));
+
+        ArrayList<Component> allComponents = componentList.findAllComponents();
+
+        if (studentGrades.isEmpty()) {
+            throw new StudentNotFoundException("No grades found for student: " + matricNumber);
+        }
+
+        boolean hasAllComponents = allComponents.stream()
+                .allMatch(component -> studentGrades.stream()
+                        .anyMatch(grade -> grade.getComponent().equals(component)));
+
+        if (!hasAllComponents) {
+            throw new IncompleteGradesException(
+                    "Cannot calculate GPA: Student " + matricNumber + " is missing grades for some components");
+        }
+
+        return studentGrades
+                .stream()
+                .mapToDouble(grade -> grade.getScore() * grade.getComponent().getWeight())
+                .sum();
+    }
```

Comment:
Method is doing too many things (SLAP violation) - consider breaking into smaller methods. Magic number 0.07 in tax calculation could be a constant